Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs

```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
 	/// </summary>
 	[Parameter] public RenderFragment InputGroupEndTemplate { get; set; }
 
+	/// <summary>
+	/// Enables filtering capabilities.
+	/// </summary>
+	[Parameter] public bool AllowFiltering { get; set; }
+
+	[Parameter] public Func<TItem, string, bool> FilterPredicate { get; set; }
+
+	[Parameter] public bool ClearFilterOnHide { get; set; }
+
+	/// <summary>
+	/// This event is fired when a dropdown element has been made visible to the user.
+	/// </summary>
+	[Parameter] public EventCallback<string> OnShown { get; set; }
+
+	/// <summary>
+	/// This event is fired when a dropdown element has been hidden from the user.
+	/// </summary>
+	[Parameter] public EventCallback<string> OnHidden { get; set; }
+
+	/// <summary>
+	/// Template that defines what should be rendered in case of empty items.
+	/// </summary>
+	[Parameter] public RenderFragment EmptyTemplate { get; set; }
+
+	[Parameter] public bool AllowSelectAll { get; set; }
+
+	[Parameter] public EventCallback<bool> SelectAllChanged { get; set; }
+
+	[Parameter] public string SelectAllText { get; set; }
+
 	/// <summary>
 	/// Additional attributes to be splatted onto an underlying HTML element.
 	/// </summary>
 	[Parameter(CaptureUnmatchedValues = true)] public Dictionary<string, object> AdditionalAttributes { get; set; }
 
+	[Inject] private IJSRuntime JSRuntime { get; set; } = default!;
+
 	protected bool HasInputGroupsEffective => !String.IsNullOrWhiteSpace(InputGroupStartText) || !String.IsNullOrWhiteSpace(InputGroupEndText) || (InputGroupStartTemplate is not null) || (InputGroupEndTemplate is not null);
 
-	private ElementReference inputElement;
+	/// <summary>
+	/// Triggers the <see cref="OnHidden"/> event. Allows interception of the event in derived components.
+	/// </summary>
+	protected virtual Task InvokeOnHiddenAsync(string elementId) => OnHidden.InvokeAsync(elementId);
+
+	/// <summary>
+	/// Triggers the <see cref="OnShown"/> event. Allows interception of the event in derived components.
+	/// </summary>
+	protected virtual Task InvokeOnShownAsync(string elementId) => OnShown.InvokeAsync(elementId);
+
+	private IJSObjectReference jsModule;
+	private readonly DotNetObjectReference<HxMultiSelectInternal<TValue, TItem>> dotnetObjectReference;
+	private ElementReference elementReference;
+	private ElementReference filterInputReference;
+	private bool isShown;
+	private string filterText = string.Empty;
+	private bool disposed;
 
-	private async Task HandleItemSelectionChangedAsync(bool newChecked, TItem item)
+	public HxMultiSelectInternal()
+	{
+		dotnetObjectReference = DotNetObjectReference.Create(this);
+	}
+
+	/// <inheritdoc cref="ComponentBase.OnAfterRenderAsync(bool)" />
+	protected override async Task OnAfterRenderAsync(bool firstRender)
+	{
+		await base.OnAfterRenderAsync(firstRender);
```

Comment:
No need to call empty `base` method. We are in a library code where we can be a little bit more aggressive.